### ISSUE TITLE
`Inf` is treated as variable rather than data

### ISSUE THREAD

**Issue by xukai92** [TuringLang/Turing.jl#912]
In the MWE below, `Inf` is treated as a variable rather than the literal `Inf`.

```julia
using Turing

@model mwe() = begin
    m ~ Normal(0, 1)
    Inf ~ Normal(m, 1)
end

sample(mwe(), HMC(0.1, 2), 100)
```

**Comment by xukai92** [comment]
Reported by @jonasmac16.

**Comment by mohamed82008** [comment]
This is an interesting one, but may I ask how is support `Inf` as data useful?

**Comment by xukai92** [comment]
It's for something like below

```julia
res = solve(...)
if res.success
    data ~ dist
else
    Inf ~ dist
end
```

**Comment by mohamed82008** [comment]
Can we achieve the same goal by defining the `Impossible` distribution? `rand(::Impossible) = NaN` and `logpdf(::Impossible, ::Any) = -Inf`.
```julia
res = solve(...)
if res.success
    data ~ dist
else
    data ~ Impossible()
end
```
Supporting `Inf` on the LHS is also possible.

**Comment by xukai92** [comment]
Yes there are maily work aournds for this. Like the one you post or just pass in a data call `data_inf` which is `Inf`. But it's might be cool for the user to do so easily.

**Comment by yebai** [comment]
Maybe standardize the compiler API for LHS with concrete values, by e.g. introducing a function operate on Julia expressions (which can be called by the `@model` macro)? This would allow the user to overload `~` for specific use cases.

**Comment by mohamed82008** [comment]
Sounds good.
